File: PdfClown-NetStandard/CliSamples/org/pdfclown/samples/cli/ImageExtractionSample.cs
using org.pdfclown.bytes;
using org.pdfclown.documents;
using files = org.pdfclown.files;
using org.pdfclown.objects;

using System;
using System.IO;

namespace org.pdfclown.samples.cli
{
  /**
    <summary>This sample demonstrates how to extract XObject images from a PDF document.</summary>
    <remarks>
      <para>Inline images are ignored.</para>
      <para>XObject images other than JPEG aren't currently supported for handling.</para>
    </remarks>
  */
  public class ImageExtractionSample
    : Sample
  {
    public override void Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptFileChoice("Please select a PDF file");
      using(var file = new files::File(filePath))
      {
        // 2. Iterating through the indirect object collection...
        int index = 0;
        foreach(PdfIndirectObject indirectObject in file.IndirectObjects)
        {
          // Get the data object associated to the indirect object!
          PdfDataObject dataObject = indirectObject.DataObject;
          // Is this data object a stream?
          if(dataObject is PdfStream)
          {
            PdfDictionary header = ((PdfStream)dataObject).Header;
            // Is this stream an image?
            if(header.ContainsKey(PdfName.Type)
              && header[PdfName.Type].Equals(PdfName.XObject)
              && header[PdfName.Subtype].Equals(PdfName.Image))
            {
              // Which kind of image?
              if(header[PdfName.Filter].Equals(PdfName.DCTDecode)) // JPEG image.
              {
                // Get the image data (keeping it encoded)!
                IBuffer body = ((PdfStream)dataObject).GetBody(false);
                // Export the image!
                ExportImage(
                  body,
                  "ImageExtractionSample_" + (index++) + ".jpg"
                  );
              }
              else // Unsupported image.
              {Console.WriteLine("Image XObject " + indirectObject.Reference + " couldn't be extracted (filter: " + header[PdfName.Filter] + ")");}
            }
          }
        }
      }
    }

    private void ExportImage(
      IBuffer data,
      string filename
      )
    {
      string outputPath = GetOutputPath(filename);
      FileStream outputStream;
      try
      {outputStream = new FileStream(outputPath, FileMode.CreateNew);}
      catch(Exception e)
      {throw new Exception(outputPath + " file couldn't be created.",e);}

      try
      {
        BinaryWriter writer = new BinaryWriter(outputStream);
        writer.Write(data.ToByteArray());
        writer.Close();
        outputStream.Close();
      }
      catch(Exception e)
      {throw new Exception(outputPath + " file writing has failed.",e);}

      Console.WriteLine("Output: " + outputPath);
    }
  }
}

File: java/pdfclown.samples.cli/src/org/pdfclown/samples/cli/ImageExtractionSample.java
package org.pdfclown.samples.cli;

import org.pdfclown.bytes.IBuffer;
import org.pdfclown.files.File;
import org.pdfclown.objects.PdfDataObject;
import org.pdfclown.objects.PdfDictionary;
import org.pdfclown.objects.PdfIndirectObject;
import org.pdfclown.objects.PdfName;
import org.pdfclown.objects.PdfStream;
import org.pdfclown.util.io.IOUtils;

/**
  This sample demonstrates <b>how to extract XObject images</b> from a PDF document.
  <h3>Remarks</h3>
  <p>Inline images are ignored.</p>
  <p>XObject images other than JPEG aren't currently supported for handling.</p>

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @version 0.1.2.1, 04/08/15
*/
public class ImageExtractionSample
  extends Sample
{
  @Override
  public void run(
    )
  {
    File file = null;
    try
    {
      // 1. Opening the PDF file...
      {
        String filePath = promptFileChoice("Please select a PDF file");
        try
        {file = new File(filePath);}
        catch(Exception e)
        {throw new RuntimeException(filePath + " file access error.",e);}
      }

      // 2. Iterating through the indirect object collection...
      int index = 0;
      for(PdfIndirectObject indirectObject : file.getIndirectObjects())
      {
        // Get the data object associated to the indirect object!
        PdfDataObject dataObject = indirectObject.getDataObject();
        // Is this data object a stream?
        if(dataObject instanceof PdfStream)
        {
          PdfDictionary header = ((PdfStream)dataObject).getHeader();
          // Is this stream an image?
          if(header.containsKey(PdfName.Type)
            && header.get(PdfName.Type).equals(PdfName.XObject)
            && header.get(PdfName.Subtype).equals(PdfName.Image))
          {
            // Which kind of image?
            if(header.get(PdfName.Filter).equals(PdfName.DCTDecode)) // JPEG image.
            {
              // Get the image data (keeping it encoded)!
              IBuffer body = ((PdfStream)dataObject).getBody(false);
              // Export the image!
              exportImage(
                body,
                "ImageExtractionSample_" + (index++) + ".jpg"
                );
            }
            else // Unsupported image.
            {System.out.println("Image XObject " + indirectObject.getReference() + " couldn't be extracted (filter: " + header.get(PdfName.Filter) + ")");}
          }
        }
      }
    }
    finally
    {
      // 3. Closing the PDF file...
      IOUtils.closeQuietly(file);
    }
  }

  private void exportImage(
    IBuffer data,
    String filename
    )
  {
    java.io.File outputFile = new java.io.File(getOutputPath(filename));
    java.io.BufferedOutputStream outputStream;
    try
    {
      outputFile.createNewFile();
      outputStream = new java.io.BufferedOutputStream(
        new java.io.FileOutputStream(outputFile)
        );
    }
    catch(Exception e)
    {throw new RuntimeException(outputFile.getPath() + " file couldn't be created.",e);}

    try
    {outputStream.write(data.toByteArray());}
    catch(Exception e)
    {throw new RuntimeException(outputFile.getPath() + " file writing has failed.",e);}
    finally
    {IOUtils.close(outputStream);}

    System.out.println("Output: " + outputFile.getPath());
  }
}